ArduCopter/ekf_check.cpp

```diff
@@ -72,8 +77,10 @@ void Copter::ekf_check()
             // if compass is flagged as bad and the counter reaches zero then clear flag
             if (ekf_check_state.bad_variance && ekf_check_state.fail_count == 0) {
                 ekf_check_state.bad_variance = false;
-                // log recovery in the dataflash
-                Log_Write_Error(ERROR_SUBSYSTEM_EKFCHECK, ERROR_CODE_EKFCHECK_VARIANCE_CLEARED);
+                if (should_act) {
+                    // log recovery in the dataflash
+                    Log_Write_Error(ERROR_SUBSYSTEM_EKFCHECK, ERROR_CODE_EKFCHECK_VARIANCE_CLEARED);
+                }
                 // clear failsafe
                 failsafe_ekf_off_event();
```

Comment:
Humm, good point. I said I wasn't 100% convinced on the right solution, here I am tending to the other :smile: 

The scenario I was thinking to be in favor of the second option was you would have an EKF failsafe in the air, you land, disarm, it clears but then wouldn't be a message saying it was cleared (which will only happens with log disarmed enabled but still).

I should say that there are two log messages in play here:

- ERROR_SUBSYSTEM_**FAILSAFE_EKFINAV**, FAILSAFE_OCCURRED / FAILSAFE_RESOLVED
- ERROR_SUBSYSTEM_**EKFCHECK**, EKFCHECK_BAD_VARIANCE / EKFCHECK_VARIANCE_CLEARED

So I'm thinking that maybe the right approach is:

- always write the ERROR_SUBSYSTEM_**EKFCHECK** one
- only call *failsafe_ekf_event* if *should_act*
- always call *failsafe_ekf_off_event* (which will return immediately if there was no failsafe active)

The other solution is to rewrite these entire methods and simplify all of it.